Correct Variant::hash_compare()

hpvb opened the pull request:
There was a logic error in #7815 which made
Variant.hash_compare() == Variant.hash_compare() always true.
In an attempt to short-circuit the NaN check I made an (in hindsight) obvious
error: 10 == 12 || is_nan(10) == is_nan(12)

This will be true for all inputs, except for the NaN, not-NaN case. The macro
has been updated to now generate:

(10 == 12) || (is_nan(10) && is_nan(10))

so:

(10 == 12)   || (is_nan(10)  && is_nan(12))  = false
   False  or (False and False) is False
(10 == 10)   || (is_nan(10)  && is_nan(10))  = true
   True or (False and False) is True
(Nan == 10)  || (is_nan(NaN) && is_nan(10))  = false
   False or (True and False) is False
(Nan == Nan) || (is_nan(NaN) && is_nan(NaN)) = true
   False or (True and True) is True

Which is correct for all cases. 

This bug was triggered because the hash function for floating point numbers
can very easily generate collisions for the tested Vector3(). I've also added
an extra hashing step to the float hash function to make this less likely to
occur.

This fixes #8081 and probably many more random weirdness.

bojidar-bg commented:
Still fails I think:

A | B | Wanted | Actual | Comment
-- | - | -------- | ------ | -------
NAN | NAN | True | True | First `if` entered, second `if` entered, broke out.
NAN | 0 | False | False | First `if` entered, second `if` not entered, returning `false`.
0 | NAN | False | False | First `if` entered, second `if` not entered, returning `false`.
0 | 0 | True | True | First `if` not entered, exiting loop.
0 | 1 | False | **True** | First `if` not entered, **second `if` entered, broke out.**

Reason: Code is equivalent to:
```cpp
if ((p_lhs) != (p_rhs) && Math::is_nan(p_lhs) != Math::is_nan(p_rhs))
        return false;
```

Maybe this would be better?

```cpp
if (p_lhs == p_rhs) {
        return true;
} else if (Math::is_nan(p_lhs) && Math::is_nan(p_rhs)) {
        return true;
} else {
        return false;
}
```
Which is equivalent to:
```cpp
if (p_lhs == p_rhs || (Math::is_nan(p_lhs) && Math::is_nan(p_rhs))) {
        return true;
} else {
        return false;
}
```
Which is equivalent to:
```cpp
return p_lhs == p_rhs || (Math::is_nan(p_lhs) && Math::is_nan(p_rhs));
```

A | B | Wanted | Actual | Comment
-- | - | -------- | ------ | -------
NAN | NAN | True | True | `is_nan(p_lhs) && is_nan(p_rhs)` is `true`.
NAN | 0 | False | False | Both `p_lhs == p_rhs` and `is_nan(p_lhs) && is_nan(p_rhs)` are `false`.
0 | NAN | False | False | Both `p_lhs == p_rhs` and `is_nan(p_lhs) && is_nan(p_rhs)` are `false`.
0 | 0 | True | True | `p_lhs == p_rhs` is `true`.
0 | 1 | False | *False* | `p_lhs == p_rhs` is `false` and `is_nan(p_lhs)` is `false` too.

hpvb commented:
Urgh, I need to sleep more. The nan check should just be is_nan && is_nan. I'll revise..

hpvb committed:
Correct Variant::hash_compare()

There was a logic error in #7815 which made
Variant.hash_compare() == Variant.hash_compare() always true.
In an attempt to short-circuit the NaN check I made an (in hindsight) obvious
error: 10 == 12 || is_nan(10) == is_nan(12)

This will be true for all inputs, except for the NaN, not-NaN case. The macro
has been updated to now generate:

(10 == 12) || (is_nan(10) && is_nan(10))

so:

(10 == 12)   || (is_nan(10)  && is_nan(12))  = false
   False  or (False and False) is False
(10 == 10)   || (is_nan(10)  && is_nan(10))  = true
   True or (False and False) is True
(Nan == 10)  || (is_nan(NaN) && is_nan(10))  = false
   False or (True and False) is False
(Nan == Nan) || (is_nan(NaN) && is_nan(NaN)) = true
   False or (True and True) is True

Which is correct for all cases.

This bug was triggered because the hash function for floating point numbers
can very easily generate collisions for the tested Vector3(). I've also added
an extra hashing step to the float hash function to make this less likely to
occur.

This fixes #8081 and probably many more random weirdness.

hpvb commented:
I have corrected the PR, please go over it again in case having additional sleep did not work as well as I had hoped :) Thanks so much for catching this @bojidar-bg

bojidar-bg reviewed (approved):
Looks good; works good
